core/src/murmur.rs

```diff
@@ -0,0 +1,351 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+//! The murmur protocol implementation
+//!
+use alloc::{
+    string::String, vec, 
+    vec::Vec,
+    collections::BTreeMap,
+};
+
+#[cfg(feature = "client")]
+use crate::otp::BOTPGenerator;
+
+#[cfg(feature = "client")]
+use rand_core::OsRng;
+
+use crate::types::*;
+use etf_crypto_primitives::{
+    ibe::fullident::Identity,
+    encryption::tlock::*
+};
+use w3f_bls::{DoublePublicKey, EngineBLS};
+use ckb_merkle_mountain_range::{
+    MerkleProof,
+    util::{
+        MemMMR,
+        MemStore,
+    },
+    MMRStore,
+};
+use ark_serialize::CanonicalSerialize;
+use codec::Encode;
+use sha3::Digest;
+
+#[derive(Debug)]
+pub enum Error {
+    ExecuteError,
+    MMRError,
+}
+
+/// the murmur store
+#[cfg(feature = "client")]
+#[derive(Clone, serde::Serialize, serde::Deserialize)]
+pub struct MurmurStore {
+    /// map block numbers to positions in the mmr
+    pub metadata: BTreeMap<BlockNumber, Ciphertext>,
+    /// the root of the mmr
+    pub root: Leaf,
+}
+
+#[cfg(feature = "client")]
+impl MurmurStore {
+
+    /// creates the leaves needed to generate an MMR
+    /// This function generates otp codes for the given block schedule
+    /// then it encrypts the resulting codes and constructs leaves 
+    /// the leaves can be used to generate an MMR
+    ///
+    pub fn new<E: EngineBLS, I: IdentityBuilder<BlockNumber>>(
+        seed: Vec<u8>,
+        block_schedule: Vec<BlockNumber>,
+        ephemeral_msk: [u8;32],
+        pk: DoublePublicKey<E>,
+    ) -> Self {
+        let totp = build_generator(&seed.clone());
+        let mut metadata = BTreeMap::new();
+
+        let store = MemStore::default();
+        let mut mmr = MemMMR::<_, MergeLeaves>::new(0, store);
+        
+        for i in &block_schedule {
+            let otp_code = totp.generate(*i);
+            let identity = I::build_identity(*i);
+            let ct_bytes = timelock_encrypt::<E>(
+                identity,
+                pk.1,
+                ephemeral_msk,
+                otp_code.as_bytes(),
+            );
+            let leaf = Leaf(ct_bytes.clone());
+            let _pos = mmr.push(leaf).expect("todo");//.map_err(|e| {
+            metadata.insert(*i, ct_bytes);
+        }
+
+       
+        MurmurStore {
+            metadata,
+            root: mmr.get_root().unwrap().clone(),
+        }
+    }
+
+    /// build the parameters needed to use a murmur wallet
+    /// note: this rebuilds the entire mmr
+    /// we can look into ways to optimize this in the future
+    /// the main issue is that he MemStore is not serializable
+    /// a possible fix is to externalize mmr logic
+    ///
+    /// TODO: this should probably be a result, not option
```

Comment:
```suggestion
    /// Build the parameters needed to use a murmur wallet
    /// note: this rebuilds the entire mmr
    /// we can look into ways to optimize this in the future
    /// the main issue is that he MemStore is not serializable
    /// a possible fix is to externalize mmr logic
    // TODO: this should probably be a result, not option
```

Do we have a task for the TODO ?